File: regions/system/library/dart/kafka/lib/src/topic.dart
import 'error.dart';

final _topicNameRegex = RegExp(
    r'^k1s0\.(system|business|service)\.[a-z0-9-]+\.[a-z0-9-]+\.v[0-9]+$');

class TopicConfig {
  final String name;
  final int? partitions;
  final int? replicationFactor;
  final int? retentionMs;

  const TopicConfig({
    required this.name,
    this.partitions,
    this.replicationFactor,
    this.retentionMs,
  });

  void validateName() {
    if (name.isEmpty) {
      throw KafkaError('topic name must not be empty');
    }
    if (!_topicNameRegex.hasMatch(name)) {
      throw KafkaError('invalid topic name: $name');
    }
  }

  /// トピック設定の全フィールドを検証する。
  /// L-004 監査対応: name に加えて partitions/replicationFactor/retentionMs の正値チェックを追加する。
  /// null 値は「未指定」として許容する（ブローカーデフォルトを使用する意図）。
  void validate() {
    // トピック名の検証を先に実行する
    validateName();
    // パーティション数が指定されている場合は正の整数であることを検証する
    if (partitions != null && partitions! <= 0) {
      throw ArgumentError('partitions は正の整数でなければなりません: $partitions');
    }
    // レプリケーションファクターが指定されている場合は正の整数であることを検証する
    if (replicationFactor != null && replicationFactor! <= 0) {
      throw ArgumentError(
          'replicationFactor は正の整数でなければなりません: $replicationFactor');
    }
    // 保持時間（ms）が指定されている場合は正の整数であることを検証する
    if (retentionMs != null && retentionMs! <= 0) {
      throw ArgumentError('retentionMs は正の整数でなければなりません: $retentionMs');
    }
  }

  String tier() {
    final match = _topicNameRegex.firstMatch(name);
    return match?.group(1) ?? '';
  }

  /// tier 別デフォルトパーティション数を返す。
  int defaultPartitionsForTier() {
    return TopicConfig.partitionsForTier(tier());
  }

  /// tier 名からデフォルトパーティション数を返す。
  ///
  /// - system tier: 6 パーティション
  /// - business tier: 6 パーティション
  /// - service tier / その他: 3 パーティション
  static int partitionsForTier(String tier) {
    switch (tier) {
      case 'system':
      case 'business':
        return 6;
      default:
        return 3;
    }
  }
}

File: regions/system/library/go/kafka/topic.go
package kafka

import (
	"fmt"
	"regexp"
)

// トピック名のバリデーション正規表現
// パターン: k1s0.(system|business|service).<domain>.<event>.v<version>
var topicNameRegex = regexp.MustCompile(`^k1s0\.(system|business|service)\.[a-z0-9-]+\.[a-z0-9-]+\.v[0-9]+$`)

// TopicPartitionInfo はトピックのパーティション情報を表す。
type TopicPartitionInfo struct {
	// Topic はトピック名。
	Topic string
	// Partition はパーティション番号。
	Partition int32
	// Leader はリーダーブローカーの ID。
	Leader int32
	// Replicas はレプリカが配置されているブローカー ID のリスト。
	Replicas []int32
	// ISR (In-Sync Replicas) は同期しているレプリカのブローカー ID のリスト。
	ISR []int32
}

// TopicConfig は Kafka トピック設定。
type TopicConfig struct {
	// Name はトピック名。
	Name string
	// Partitions はパーティション数。
	Partitions int
	// ReplicationFactor はレプリケーション係数。
	ReplicationFactor int
	// RetentionMs はメッセージ保持期間 (ミリ秒)。
	RetentionMs int64
}

// ValidateName はトピック名を検証する。
// 正しい形式: k1s0.(system|business|service).<domain>.<event>.v<version>
func (t *TopicConfig) ValidateName() error {
	if t.Name == "" {
		return fmt.Errorf("topic name must not be empty")
	}
	if !topicNameRegex.MatchString(t.Name) {
		return fmt.Errorf("invalid topic name: %s (expected format: k1s0.(system|business|service).<domain>.<event>.v<version>)", t.Name)
	}
	return nil
}

// Tier はトピック名からティアを返す (system, business, service)。
func (t *TopicConfig) Tier() string {
	matches := topicNameRegex.FindStringSubmatch(t.Name)
	if len(matches) < 2 {
		return ""
	}
	return matches[1]
}

// DefaultPartitionsForTier は tier 別のデフォルトパーティション数を返す。
//   - system tier: 6 パーティション
//   - business tier: 6 パーティション
//   - service tier / その他: 3 パーティション
func DefaultPartitionsForTier(tier string) int {
	switch tier {
	case "system", "business":
		return 6
	default:
		return 3
	}
}

// WithTierDefaults はトピック名から tier を判定し、
// tier 別デフォルトパーティション数を設定して返す。
func (t *TopicConfig) WithTierDefaults() *TopicConfig {
	tier := t.Tier()
	if tier != "" {
		t.Partitions = DefaultPartitionsForTier(tier)
	}
	return t
}
